sync_adapter_sink.go

```diff
@@ -57,7 +54,10 @@ func (spa *synchronousMessageSinkAdapter) loop() {
 	eg, ctx := errgroup.WithContext(context.Background())
 
 	eg.Go(func() error {
-		return spa.aprod.PublishMessages(ctx, acks, toSend)
+		if err := spa.aprod.PublishMessages(ctx, acks, toSend); err != nil {
+			return err
+		}
+		return errSinkClosed // return error to shutdown the main loop
```

Comment:
Using an error to effectively control flow seems unpleasant.  How about a context.WithCancel() higher up and we can just cancel it here?

Reply:
yeah, that would be nicer, will change it to do that

Reply:
looks much better now 👍